Blish HUD/Controls/TextInputBase.cs

```diff
@@ -0,0 +1,718 @@
+ *
+ *  MIT License
+ *
+ *  Copyright (c) 2017-2020 The Myra Team
+ *
+ *  Permission is hereby granted, free of charge, to any person obtaining a copy
+ *  of this software and associated documentation files (the "Software"), to deal
+ *  in the Software without restriction, including without limitation the rights
+ *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *  copies of the Software, and to permit persons to whom the Software is
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in all
+ *  copies or substantial portions of the Software.
+
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Blish_HUD.Input;
+using Microsoft.Xna.Framework;
+using Microsoft.Xna.Framework.Graphics;
+using MonoGame.Extended.BitmapFonts;
+using Microsoft.Xna.Framework.Input;
+
+namespace Blish_HUD.Controls {
+
+    /// <summary>
+    /// The base of controls such as the <see cref="TextBox"/> or <see cref="MultilineTextBox"/> which accept text input from the user.
+    /// </summary>
+    public abstract class TextInputBase : Control {
+
+        private static readonly Logger Logger = Logger.GetLogger<TextInputBase>();
+
+        protected static readonly char[] WordSeperators = { ' ', '\n', '`', '~', '!', '@', '#', '%', '^', '&', '*', '(', ')', '-', '=', '+', '[', '{', ']', '}', '\\', '|', ';', ':', '\'', '"', ',', '.', '<', '>', '/', '?' };
+
+        #region Load Static
+
+        private static readonly Color _highlightColor;
+
+        static TextInputBase() {
+            _highlightColor = new Color(92, 80, 103, 150);
+        }
+
+        #endregion
+
+        protected const char NEWLINE = '\n';
+
+        /// <summary>
+        /// Fires when the <see cref="Text"/> is changed.
+        /// </summary>
+        public event EventHandler<EventArgs> TextChanged;
+
+        /// <summary>
+        /// Fires when the <see cref="CursorIndex"/> is changed.
+        /// </summary>
+        public event EventHandler<ValueEventArgs<int>> CursorIndexChanged;
+
+        protected void OnTextChanged(ValueChangedEventArgs<string> e) => TextChanged?.Invoke(this, e);
+
+        protected void OnCursorIndexChanged(ValueEventArgs<int> e) {
+            _cursorMoved = true;
+
+            UpdateScrolling();
```

Comment:
Why not letting the implementation classes either subscribing to the `CursorIndexChanged`, or let them override `OnCurserIndexChanged`?
With this the `MultilineTextBox` needs to "implement" the `UpdateScrolling` w/o even doing anything in there

Reply:
`MultilineTextBox` will actually need to implement `UpdateScrolling` at some point, but the exact behavior hasn't been decided on.

Reply:
Copied from when this was originally brought up.  It'll get handled in a separate PR, but something to look over in case you have thoughts.

> - Decide on how to support text that extends beyond the width of the control (word wrap?  horizontal scroll bar? something else?) - I am not looking forward to implementing whatever is decided here.  The game wraps lines that extend beyond the end of the control.
> - Decide on how to support text that extends beyond the bottom of the control (vertical scroll bar? something else?) - The game lets this happen, but does not show a vertical scrollbar, so scrolling up and down when this happens is weird.

Reply:
I've no problem if no one will not implement it in the future, so can be resolved